pkg/controller/machinehealthcheck/machinehealthcheck_controller.go

```diff
@@ -271,38 +295,32 @@ func (r *ReconcileMachineHealthCheck) getMachinesFromMHC(mhc mhcv1beta1.MachineH
 }
 
 func (r *ReconcileMachineHealthCheck) getMachineFromNode(node corev1.Node) (*mapiv1.Machine, error) {
-	machineKey, ok := node.Annotations[machineAnnotationKey]
-	if !ok {
-		glog.V(4).Infof("no machine annotation for node %s", node.GetName())
-		return nil, nil
-	}
-	glog.V(4).Infof("Node %s is annotated with machine %s", node.GetName(), machineKey)
-
-	namespace, machineName, err := cache.SplitMetaNamespaceKey(machineKey)
-	if err != nil {
-		return nil, fmt.Errorf("machine name has wrong format %v", machineKey)
-	}
-	machine := &mapiv1.Machine{}
-	if err = r.client.Get(
+	machineList := &mapiv1.MachineList{}
+	if err := r.client.List(
 		context.TODO(),
-		types.NamespacedName{
-			Namespace: namespace,
-			Name:      machineName,
-		},
-		machine,
+		machineList,
+		client.MatchingFields{machineNodeNameIndex: node.Name},
 	); err != nil {
-		return nil, fmt.Errorf("error getting machine: %v", err)
+		return nil, fmt.Errorf("failed getting machine list: %v", err)
 	}
-	return machine, nil
+	if len(machineList.Items) != 1 {
+		return nil, fmt.Errorf("expecting one machine for node %v, got: %v", node, machineList.Items)
+	}
+	return &machineList.Items[0], nil
 }
 
 func (r *ReconcileMachineHealthCheck) mhcRequestsFromNode(o handler.MapObject) []reconcile.Request {
 	glog.V(4).Infof("Getting MHC requests from node %q", namespacedName(o.Meta).String())
 	node := &corev1.Node{}
 	if err := r.client.Get(context.Background(), namespacedName(o.Meta), node); err != nil {
-		glog.Errorf("No-op: Unable to retrieve node %q from store: %v", namespacedName(o.Meta).String(), err)
-		return nil
+		if apimachineryerrors.IsNotFound(err) {
+			node.Name = o.Meta.GetName()
```

Comment:
If we only ever care about the name, we should maybe just change `getMachineFromNode()` to take the name as a parameter instead of the full object.